modules/ray_tracing/test/src/userobjects/LotsOfRaysRayStudy.C

```diff
@@ -0,0 +1,282 @@
+  // it is rectangular)
+  if (_compute_expected_distance && !isRectangularDomain())
+    paramError("compute_expected_distance", "Not supported when the domain is non-rectangular.");
+
+  const bool polar_set = parameters.isParamSetByUser("polar_quad_order");
+  const bool azimuthal_set = parameters.isParamSetByUser("azimuthal_quad_order");
+  if (_mesh.dimension() == 1)
+  {
+    if (polar_set)
+      paramError("polar_quad_order", "Not needed for 1D");
+    if (azimuthal_set)
+      paramError("azimuthal_quad_order", "Not needed for 1D");
+  }
+  if (!_side_aq && !_centroid_aq)
+  {
+    if (polar_set)
+      paramError("polar_quad_order", "Not needed without side_aq or centroid_aq enabled");
+    if (azimuthal_set)
+      paramError("azimuthal_quad_order", "Not needed without side_aq or centroid_aq enabled");
+  }
+  if (azimuthal_set && _mesh.dimension() == 2)
+    paramError("azimuthal_quad_order", "Not needed for 2D");
+}
+
+void
+LotsOfRaysRayStudy::defineRays()
+{
+  if (_compute_expected_distance)
+  {
+    _bbox_intersection_helper =
+        libmesh_make_unique<BoundingBoxIntersectionHelper>(boundingBox(), _mesh.dimension());
+    _expected_distance = 0;
+  }
+
+  // Gather the boundary elem centroids
+  std::vector<Point> boundary_centroids;
+  if (_centroid_to_centroid)
+  {
+    for (const auto & elem : *_mesh.getActiveLocalElementRange())
+      if (elem->on_boundary())
+        boundary_centroids.push_back(elem->centroid());
+    _comm.allgather(boundary_centroids);
+  }
+
+  std::vector<std::pair<Real, Real>> half_aq_angles;
+  if (_side_aq || _centroid_aq)
+  {
+    std::vector<Real> unused_weights;
+    if (_mesh.dimension() == 2)
+      RayTracingAngularQuadrature::getHalfRange2D(
+          _polar_quad_order, half_aq_angles, unused_weights, 0.5 * M_PI);
+    else
+      RayTracingAngularQuadrature::getHalfRange3D(
+          4 * _azimuthal_quad_order, _polar_quad_order, half_aq_angles, unused_weights, 0.5 * M_PI);
+  }
+
+  for (const auto & bnd_elem : *_mesh.getBoundaryElementRange())
+  {
+    const auto side = bnd_elem->_side;
+    const Elem * elem = bnd_elem->_elem;
+
+    // Starting from local elems only
+    if (elem->processor_id() != _pid)
+      continue;
+
+    // Centroid on this boundary face
+    const auto & side_centroid = sidePtrHelper(elem, side)->centroid();
+
+    // Vertices on this boundary side -> in direction of all other vertices not on said boundary
+    // side on elem
+    if (_vertex_to_vertex)
+      for (const auto n : elem->nodes_on_side(side))
+        if (elem->is_vertex(n))
+          for (unsigned int v_to = 0; v_to < elem->n_vertices(); ++v_to)
+            if (!elem->is_node_on_side(v_to, side))
+              defineRay(elem, side, elem->point(n), elem->point(v_to), false);
+
+    // Centroid on this boundary side -> in direction of all other vertices not on said boundary
+    // side on elem
```

Comment:
why not on said boundary?  Is the tangeant case captured in another test?

Reply:
Yep